File: app/src/main/java/io/github/benoitduffez/cupsprint/UntrustedCertActivity.java
package io.github.benoitduffez.cupsprint;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.jonbanjo.ssl.JfSSLScheme;

import java.security.cert.X509Certificate;

/**
 * Show an untrusted cert info + two buttons to accept or refuse to trust said cert
 */
public class UntrustedCertActivity extends Activity {
    public static final String KEY_CERT = UntrustedCertActivity.class.getName() + ".Certs";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.untrusted_cert);

        final X509Certificate cert = (X509Certificate) getIntent().getSerializableExtra(KEY_CERT);

        // Build short cert description
        StringBuilder sb = new StringBuilder();
        sb.append("Issuer: ").append(cert.getIssuerX500Principal().toString());
        sb.append("\nValidity: not before ").append(cert.getNotBefore().toString());
        sb.append("\nValidity: not after ").append(cert.getNotAfter().toString());
        sb.append("\nSubject: ").append(cert.getSubjectX500Principal().getName());
        sb.append("\nKey algo: ").append(cert.getSigAlgName());
        sb.append("\nKey: ").append(cert.getPublicKey().toString());

        TextView certInfo = (TextView) findViewById(R.id.untrusted_certinfo);
        certInfo.setText(sb);

        findViewById(R.id.untrusted_trust_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (JfSSLScheme.saveCertificates(new X509Certificate[]{cert})) {
                    Toast.makeText(UntrustedCertActivity.this, R.string.untrusted_trusted, Toast.LENGTH_LONG).show();
                } else {
                    Toast.makeText(UntrustedCertActivity.this, R.string.untrusted_couldnt_trust, Toast.LENGTH_LONG).show();
                }
                finish();
            }
        });

        findViewById(R.id.untrusted_abort_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });
    }
}

File: app/src/main/java/io/github/benoitduffez/cupsprint/app/UntrustedCertActivity.kt
package io.github.benoitduffez.cupsprint.app

import android.app.Activity
import android.os.Bundle
import android.view.ViewGroup
import android.widget.Toast
import io.github.benoitduffez.cupsprint.HttpConnectionManagement
import io.github.benoitduffez.cupsprint.R
import kotlinx.android.synthetic.main.untrusted_cert.*
import java.security.cert.X509Certificate

/**
 * Show an untrusted cert info + two buttons to accept or refuse to trust said cert
 */
class UntrustedCertActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.untrusted_cert)
        window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)

        val cert = intent.getSerializableExtra(KEY_CERT) as X509Certificate

        // Build short cert description
        val sb = StringBuilder()
        sb.append("Issuer: ").append(cert.issuerX500Principal.toString())
        sb.append("\nValidity: not before ").append(cert.notBefore.toString())
        sb.append("\nValidity: not after ").append(cert.notAfter.toString())
        sb.append("\nSubject: ").append(cert.subjectX500Principal.name)
        sb.append("\nKey algo: ").append(cert.sigAlgName)
        untrusted_certinfo.text = sb

        untrusted_trust_button.setOnClickListener {
            if (HttpConnectionManagement.saveCertificates(this, arrayOf(cert))) {
                Toast.makeText(this@UntrustedCertActivity, R.string.untrusted_trusted, Toast.LENGTH_LONG).show()
            } else {
                Toast.makeText(this@UntrustedCertActivity, R.string.untrusted_couldnt_trust, Toast.LENGTH_LONG).show()
            }
            finish()
        }

        untrusted_abort_button.setOnClickListener { finish() }
    }

    companion object {
        val KEY_CERT = "${UntrustedCertActivity::class.java.name}.Certs"
    }
}
